posts/react-compound-components/2021-5-25-React-compound-compnents.md

```diff
@@ -0,0 +1,250 @@
+
+You can check out the [Code sandbox link](https://codesandbox.io/s/interesting-wildflower-wj3iy) for the final demo of the accordion component.
+
+The accordion component will have four components.
+
+1. **Accordion** -  The outer wrapper component of the Accordion component. This is the root component of the Accordion component.
+2. **AccordionItem** - The component that allows us to define each accordion item. Each AccordionItem will have its AccordionButton and AccordionPanel components.
+3. **AccordionButton** - The header for the Accordion component.On clicking the accordion button will open the corresponding accordion panel.
+4. **AccordionPanel** - The panel for the accordion. This will hold the content of each accordion item.
+
+We are going to create the above mentioned components one by one and also let's see how we can create the link between them. 
+Let's start with Accordion component. Accordion component will wrap all other necessary components and will maintain the state that is to be shared among all the other components
+
+```tsx
+const Accordion: React.FC<{
+  children: ReactNode | ReactNode[];
+  className?: string
+}> = ({ children, className }) => {
+  const [activeItem, setActiveItem] = useState("");
+  // function to update the active item
+   const changeActiveItem = useCallback(
+    (value) => {
+      if (activeItem !== value) setActiveItem(value);
+    },
+    [setActiveItem, activeItem]
+  );
+
+  return <div className={className}>{children}</div>
+}
+``` 
+
+we had created the accordion component, now we need to pass or share the state and the function to update the state to its children. To achieve this, we are going to use the **React Context**. If you are not familiar with React context, please refer  [https://reactjs.org/docs/context.html](https://reactjs.org/docs/context.html). Iam not gonna explain about react context here, but i will give you a hint about react context.
+
+> Context provides a way to pass data through the component tree without having to pass props down manually at every level.
+
+Now let's add context to our accordion component.
+
+```tsx
+import { createContext, useContext } from "react";
+
+// Creating the context for the Accordion.
+export const AccordionContext = createContext<{
+  activeItem: string;
+  changeSelectedItem: (item: string) => void;
+}>({ activeItem: "", changeSelectedItem: () => {} });
+
+export const useAccordionContext = () => {
+  const context = useContext(AccordionContext);
+  if (!context) {
+    throw new Error("Error in creating the context");
+  }
+  return context;
+};
+``` 
+
+After creating the context, we need to provide values to the context, it is done by **Context.Provider** element. The values refers to the data that is to be shared among all the components in the accordion. The **Provider** component accepts a value prop to be passed to consuming components that are descendants of this Provider. One Provider can be connected to many consumers.
```

Comment:
Lets provide values to the context using the **Context.Provider** element
Remove "The values refers to the data that is to be shared among all the components in the accordion."